compiler/rustc_const_eval/src/check_consts/check.rs

```diff
@@ -584,31 +651,7 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
                     }
                 };
 
-                // Check that all trait bounds that are marked as `~const` can be satisfied.
-                //
-                // Typeck only does a "non-const" check since it operates on HIR and cannot distinguish
-                // which path expressions are getting called on and which path expressions are only used
-                // as function pointers. This is required for correctness.
-                let infcx = tcx.infer_ctxt().build(body.typing_mode(tcx));
-                let ocx = ObligationCtxt::new_with_diagnostics(&infcx);
-
-                let predicates = tcx.predicates_of(callee).instantiate(tcx, fn_args);
-                let cause = ObligationCause::new(
-                    terminator.source_info.span,
-                    self.body.source.def_id().expect_local(),
-                    ObligationCauseCode::WhereClause(callee, DUMMY_SP),
-                );
-                let normalized_predicates = ocx.normalize(&cause, param_env, predicates);
-                ocx.register_obligations(traits::predicates_for_generics(
-                    |_, _| cause.clone(),
-                    self.param_env,
-                    normalized_predicates,
-                ));
-
-                let errors = ocx.select_all_or_error();
-                if !errors.is_empty() {
-                    infcx.err_ctxt().report_fulfillment_errors(errors);
-                }
+                self.revalidate_conditional_constness(callee, fn_args, call_source, *fn_span);
```

Comment:
I am confused, why is this called outside the `if let Some(trait_did) = tcx.trait_of_item(callee) {`? Can there be non-trait conditional const calls? Given that the `const_trait_impl` feature gate is used for these calls, I assume the answer is no...

Reply:
Yes? We can put ~const bounds on free const functions.

Reply:
>  Given that the const_trait_impl feature gate is used for these calls

I think you're getting too stuck on the naming of the `const_trait_impl` feature gate. 

The feature gate enables the declaration of `const` traits, `impl const`, and the use of `~const` trait bounds on conditionally const items, which includes free const function. This is how the gate has been used for the last like... 3 years or whatever.

I don't want to rename the feature gate, since that basically. the feature gate could be called `foo_bar` for all it matters.

Reply:
That's a poorly chosen name. :(

Reply:
🤷 I'm sorry, but I didn't choose it, and I'm not really interested in doing a rename here. That can be done later, and ideally not when there are like 3 other PRs in flight concerning it.